c10/xpu/XPUFunctions.cpp

```diff
@@ -0,0 +1,188 @@
+  auto platform_list = sycl::platform::get_platforms();
+  // Enumerated GPU devices from the specific platform.
+  for (const auto& platform : platform_list) {
+    if (platform.get_backend() != sycl::backend::ext_oneapi_level_zero) {
+      continue;
+    }
+    auto device_list = platform.get_devices();
+    for (const auto& device : device_list) {
+      if (device.is_gpu()) {
+        devices.push_back(std::make_unique<sycl::device>(device));
+      }
+    }
+  }
+}
+
+static inline int deviceCountImpl(
+    std::vector<std::unique_ptr<sycl::device>>& devices) {
+  enumDevices(devices);
+  return static_cast<int>(devices.size());
+}
+
+static inline void initGlobalDevicePoolState() {
+  // Get device count and record the all GPU devices.
+  auto device_count = deviceCountImpl(gDevicePool.devices);
+  if (device_count <= 0) {
+    TORCH_WARN("XPU device count is zero!");
+    return;
+  }
+
+  // The default context is utilized for each Intel GPU device, allowing the
+  // retrieval of the context from any GPU device.
+  gDevicePool.context = std::make_unique<sycl::context>(
+      gDevicePool.devices[0]->get_platform().ext_oneapi_get_default_context());
+}
+
+static inline void initDevicePoolCallOnce() {
+  c10::call_once(init_flag, initGlobalDevicePoolState);
+}
+
+static void initDeviceProperties(DeviceProp* device_prop, int device) {
+  using namespace sycl::info;
+  using namespace sycl::ext;
+  // Get raw sycl device associated with device index.
+  auto& raw_device = *gDevicePool.devices[device];
+
+  // Initialize the device properties associated with the specific device.
+#define ASSIGN_DEVICE_PROP(property) \
+  device_prop->property = raw_device.get_info<device::property>();
+
+#define ASSIGN_EXT_DEVICE_PROP(property, default_value)                      \
+  device_prop->property = raw_device.has(sycl::aspect::ext_intel_##property) \
+      ? raw_device.get_info<intel::info::device::property>()                 \
+      : default_value;
+
+  AT_FORALL_XPU_DEVICE_PROPERTIES(ASSIGN_DEVICE_PROP);
+
+  device_prop->platform_name =
+      raw_device.get_info<device::platform>().get_info<platform::name>();
+
+  AT_FORALL_XPU_EXT_DEVICE_PROPERTIES(ASSIGN_EXT_DEVICE_PROP);
+  return;
+}
+
+static inline void check_device(int device) {
+  int total = static_cast<int>(gDevicePool.devices.size());
+  TORCH_CHECK(
+      device >= 0 && device < total,
+      "device is out of range, device is ",
+      device,
+      ", total number of device is ",
+      total,
+      ".");
+}
+
+} // anonymous namespace
+
+sycl::device& get_raw_device(int device) {
+  initDevicePoolCallOnce();
+  check_device(device);
+  return *gDevicePool.devices[device];
+}
+
+sycl::context& get_device_context() {
+  initDevicePoolCallOnce();
+  TORCH_CHECK(
+      gDevicePool.context,
+      "Device pool initialization failed, you might not have an XPU device.")
+  return *gDevicePool.context;
+}
+
+void get_device_properties(DeviceProp* device_prop, int device) {
```

Comment:
```suggestion
void get_device_properties(DeviceProp* device_prop, unsigned device) {
```

Reply:
`get_device_properties` and `get_device_idx_from_pointer` are public APIs. The other public APIs include
- `exchange_device`
- `maybe_exchange_device`
- `get_raw_device`

@malfet Regarding another backend, we use `int` in public APIs to keep consistency with its APIs. May I know do we still need to maintain this consistency? Alternatively, we use `unsigned` only in the internal API, and `int` in the public API.